integration/spadev0/main.go

```diff
@@ -0,0 +1,200 @@
+							size: perProvider[providerID].size + size,
+							replicas: append(perProvider[providerID].replicas, Replica{
+								PieceCID:        replica.PieceCID,
+								PieceLog2Size:   replica.PieceLog2Size,
+								OptionalDagRoot: replica.OptionalDagRoot,
+							}),
+						}
+					}
+				}
+
+				replicasToTest := selectReplicasToTest(perProvider)
+
+				totalCids := 0
+				totalSize := 0
+				for prov, rps := range replicasToTest {
+					provider := perProvider[prov]
+					logger.Debugf("provider %d is storing %d GiB will have %d tests\n", prov, provider.size, len(rps))
+					totalCids += len(rps)
+					totalSize += provider.size
+				}
+
+				logger.Debugf("total %d CIDs will be tested for %d providers\n", totalCids, len(replicasToTest))
+
+				AddSpadeTasks(ctx, "spadev0", replicasToTest)
+			}
+			return nil
+		},
+	}
+
+	if err := app.Run(os.Args); err != nil {
+		panic(err)
+	}
+}
+
+func fetchActiveReplicas(ctx context.Context, url string) (*ActiveReplicas, error) {
+	logger.Debugf("fetching CIDs from %s", url)
+
+	req, err := http.NewRequestWithContext(ctx,
+		http.MethodGet,
+		url,
+		nil)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create request")
+	}
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to make request")
+	}
+
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, errors.Errorf("failed to get spade CID list: %s", resp.Status)
+	}
+
+	decompressor, err := zstd.NewReader(resp.Body)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create decompressor")
+	}
+
+	defer decompressor.Close()
+
+	data, err := ioutil.ReadAll(decompressor)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to read decompressed data")
+	}
+
+	// Unmarshal the JSON data into your struct
+	var activeReplicas ActiveReplicas
+	err = json.Unmarshal(data, &activeReplicas)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to unmarshal JSON data")
+	}
+
+	return &activeReplicas, nil
+}
+
+// Compute a number of CIDs to test, based on the total size of data (assuming in GiB)
+// Minimum 1, then log2 of the size in TiB
+// ex:
+// < 4TiB = 1 cid
+// 4 TiB - 16TiB = 2 cids
+// 16 TiB - 32 TiB = 3 cids
+// 32 TiB - 64 TiB = 4 cids
+// 64 TiB - 128 TiB = 5 cids
+// 128 TiB - 256 TiB = 6 cids
+// etc...
+func numCidsToTest(size int) int {
+	return int(math.Max(math.Log2(float64(size/1024)), 1))
+}
+
+func selectReplicasToTest(perProvider map[int]ProviderReplicas) map[int][]Replica {
+	var toTest = make(map[int][]Replica)
+
+	for providerID, provider := range perProvider {
+		toTest[providerID] = make([]Replica, 0)
+
+		maxReplicas := len(provider.replicas)
+		numCidsToTest := numCidsToTest(provider.size)
+
+		// This condition should not happen, but just in case there's a situation
+		// where a massive amount of bytes are being stored in relatively few CIDs
+		if numCidsToTest > maxReplicas {
+			numCidsToTest = maxReplicas
```

Comment:
`log.Warn()` here in the unikely case your math is wrong / you hit this. E.g. id somehow `provider.replicas` is empty.